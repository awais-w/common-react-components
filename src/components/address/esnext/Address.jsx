import React from 'react';
import PropTypes from 'prop-types';
import bem from '@argos/utils/esnext/bem';
import AddressCompact from './components/AddressCompact';
import AddressDetailed from './components/AddressDetailed';
import AddressInline from './components/AddressInline';
import './Address.scss';

const cn = bem({ prefix: 'cmc-', block: 'address' });

const Address = ({ address, mode }) => {
  const result = (() => {
    switch (mode) {
      case 'INLINE':
        return <AddressInline {...address} />;
      case 'COMPACT':
        return <AddressCompact {...address} />;
      default:
        return <AddressDetailed {...address} />;
    }
  })();
  return <div className={cn()}>{result}</div>;
};

Address.propTypes = {
  address: PropTypes.shape({
    city: PropTypes.string.isRequired,
    country: PropTypes.string,
    county: PropTypes.string,
    line1: PropTypes.string.isRequired,
    line2: PropTypes.string,
    line3: PropTypes.string,
    postcode: PropTypes.string.isRequired,
  }),
  mode: PropTypes.oneOf(['INLINE', 'COMPACT', 'DETAILED']),
};
export default Address;
