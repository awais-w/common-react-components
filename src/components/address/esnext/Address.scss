@import '@cmc/common-styles/src/scss/_variables.scss';

.cmc-detailed-address {
  [class^='cmc-detailed-address'] {
    margin-bottom: $half;
  }
}

.cmc-compact-address {
  [class^='cmc-compact-address'] {
    margin-bottom: $half;
  }
}
