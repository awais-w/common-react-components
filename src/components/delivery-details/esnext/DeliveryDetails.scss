@import '@cmc/common-styles/src/scss/_variables.scss';

@mixin deliveryDetails-structure {
  display: grid;
  grid-template-columns: 1fr;
  grid-auto-columns: 1fr;
  grid-auto-flow: column;
  grid-gap: $base;
}

.cmc-delivery-details {
  @include deliveryDetails-structure;
  .cmc-icon {
    font-size: $font-size-h4;
  }
  &__contact-details {
    display: flex;
    flex-wrap: wrap;
    align-items: center;
    column-gap: $base;
    height: fit-content;
    & > div {
      display: flex;
    }
    .cmc-icon-label-value {
      color: $primary-color;
    }
  }
  &__loading > span {
    @include deliveryDetails-structure;
  }
  .cmc-inline-address {
    .cmc-icon-label-value {
      flex-wrap: nowrap;
    }
  }
}
