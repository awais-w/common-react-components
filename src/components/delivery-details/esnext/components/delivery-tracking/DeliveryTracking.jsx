import React from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';
import IconLabelValue from '@cmc/icon-label-value/esnext';
import './DeliveryTracking.scss';
import { bem } from '@argos/utils';
import DeliveryGraph from './DeliveryGraph';

const cn = bem({ prefix: 'cmc-', block: 'delivery-tracking' });

const DeliveryTracking = ({
  actualArrival,
  plannedArrival,
  instructions,
  dropNumber,
  bookedSlot,
  price,
  origin,
  isClickAndCollect,
}) => (
  <div className={cn()}>
    <div className={cn('', 'grouped')}>
      {bookedSlot && (
        <div className={cn('booked-slot')}>
          <IconLabelValue icon={''} label={'Booked slot:'} value={bookedSlot} />
        </div>
      )}
      {price && (
        <div className={cn('price')}>
          <IconLabelValue icon={''} label={'Price:'} value={price} />
        </div>
      )}
    </div>

    {(actualArrival || plannedArrival) && (
      <div className={cn('', 'grouped')}>
        {plannedArrival && (
          <div className={cn('planned-arrival')}>
            <IconLabelValue icon={''} label={'Planned:'} value={moment(plannedArrival).format('h:mma')} />
          </div>
        )}
        {actualArrival && (
          <div className={cn('actual-arrival')}>
            <IconLabelValue
              icon={''}
              label={'Arriving'}
              value={moment(actualArrival).format('h:mma')}
              isCondensed={true}
            />
          </div>
        )}
        {dropNumber && (
          <div className={cn('drop-number')}>
            <IconLabelValue icon={''} label={'Drop Number:'} value={dropNumber.toString()} />
          </div>
        )}
        {origin && (
          <div className={cn('origin')}>
            <IconLabelValue
              icon={''}
              label={'From:'}
              value={origin.name ? `${origin.name} (${origin.id})` : `${origin.id}`}
            />
          </div>
        )}
      </div>
    )}
    {instructions && (
      <div className={cn('instructions')}>
        <IconLabelValue icon={''} label={'Instructions:'} value={instructions} />
      </div>
    )}
    {!isClickAndCollect && <DeliveryGraph arrivalTime={plannedArrival} />}
  </div>
);

DeliveryTracking.propTypes = {
  bookedSlot: PropTypes.string.isRequired,
  price: PropTypes.string.isRequired,
  actualArrival: PropTypes.string,
  plannedArrival: PropTypes.string,
  dropNumber: PropTypes.number,
  origin: PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.name.isRequired,
  }),
  instructions: PropTypes.string,
};
export default DeliveryTracking;
