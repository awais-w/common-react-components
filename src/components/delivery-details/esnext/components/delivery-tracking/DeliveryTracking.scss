@import '@cmc/common-styles/src/scss/_variables.scss';
.cmc-delivery-tracking {
  @include font-standard;
  &--grouped {
    display: flex;
    margin: 0;

    div:not(:first-child) {
      margin-left: $base;
    }
  }
  &__graph {
    margin-top: $half;
  }
  .slot-graph {
    display: grid;
    grid-template-columns: auto 1fr auto 1fr auto 1fr auto;
    grid-gap: $half;
    align-items: center;
    max-width: 500px;
    margin-top: $double;
    .slot-range {
      @include font-bold;
      color: $color-sainsburys;
    }
    .slot-part {
      @include font-condensed;
      font-size: $font-size-small;
    }
    .slot-line {
      background-color: $color-sainsburys;
      height: $base;
      display: flex;
      justify-content: center;
      position: relative;
      &.first-segment {
        border-radius: $half 0 0 $half;
      }
      &.third-segment {
        border-radius: 0 $half $half 0;
      }
      svg {
        position: absolute;
        width: 20px;
        height: 14px;
        top: -18px;
      }
      &:not(.slot-active) {
        svg {
          display: none;
        }
      }
      &.slot-active {
        background-color: $color-green;
        &::before {
          content: '';
          height: $base;
          width: $base;
          background-color: $color-white;
          transform: rotate(45deg);
          margin-top: -$half;
        }
      }
    }
  }
}
