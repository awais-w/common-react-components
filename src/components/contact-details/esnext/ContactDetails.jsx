import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import IconLabelValue from '@cmc/icon-label-value/esnext';
import bem from '@argos/utils/esnext/bem';
import { formatName } from '@cmc/utils/esnext';

const cn = bem({ block: 'ContactDetails' });

const ContactDetails = ({ telephones = [], emails = [], person = {}, business, groupFields }) => {
  const telephoneCPs = telephones.map((telephone, index) => (
    <IconLabelValue
      isInline={groupFields}
      icon={groupFields && index ? '' : 'call'}
      label={telephones.length > 1 && telephone.type}
      value={telephone.number}
      key={`telephone-${telephone.number}`}
    />
  ));
  const emailCPs = emails.map((email, index) => (
    <IconLabelValue
      isInline={groupFields}
      icon={groupFields && index ? '' : 'mail'}
      label={emails.length > 1 && email.type}
      value={email.address}
      key={`email-${email.address}`}
    />
  ));

  return (
    <Fragment>
      {business && (
        <div className={cn('business')}>
          <IconLabelValue icon='business_center' value={business} />
        </div>
      )}
      <div className={cn('full-name')}>
        <IconLabelValue icon='person' value={(person.name || person.surname) && formatName({ ...person })} />
      </div>
      <div className={cn('telephones')}>{telephoneCPs}</div>
      <div className={cn('emails')}>{emailCPs}</div>
    </Fragment>
  );
};

ContactDetails.propTypes = {
  business: PropTypes.string,
  person: PropTypes.shape({
    title: PropTypes.string,
    name: PropTypes.string.isRequired,
    middleName: PropTypes.string,
    surname: PropTypes.string.isRequired,
  }).isRequired,
  emails: PropTypes.arrayOf(
    PropTypes.shape({
      type: PropTypes.string,
      address: PropTypes.string.isRequired,
    }),
  ),
  telephones: PropTypes.arrayOf(
    PropTypes.shape({
      type: PropTypes.string,
      number: PropTypes.string.isRequired,
    }),
  ),
  groupFields: PropTypes.bool,
};
export default ContactDetails;
