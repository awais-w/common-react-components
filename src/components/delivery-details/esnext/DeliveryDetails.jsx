import React from 'react';
import Skeleton from 'react-loading-skeleton';
import { bem } from '@argos/utils';
import ContactDetails from '@cmc/contact-details/esnext';
import Address from '@cmc/address/esnext';
import isEmpty from '@cmc/utils/esnext/isEmpty';
import DeliveryTracking from './components/delivery-tracking';
import propTypes from './DeliveryDetails.propTypes';
import './DeliveryDetails.scss';

const cn = bem({ prefix: 'cmc-', block: 'delivery-details' });

const DeliveryDetails = ({ contact = {}, address = {}, deliveryTracking = {}, isLoading, isClickAndCollect }) => {
  const { person, business, telephones, emails } = contact;
  const { actualArrival, plannedArrival, dropNumber, bookedSlot, price, origin, instructions } = deliveryTracking;

  return (
    <div className={cn()}>
      {!isLoading ? (
        <React.Fragment>
          {!isEmpty(contact) && (
            <div className={cn('contact-details')}>
              <ContactDetails person={person} business={business} telephones={telephones} emails={emails} groupFields />
              {Object.entries(address).length && <Address address={address} mode={'INLINE'} />}
            </div>
          )}
          <div className={cn('delivery-tracking')}>
            <DeliveryTracking
              actualArrival={actualArrival}
              plannedArrival={plannedArrival}
              dropNumber={dropNumber}
              bookedSlot={bookedSlot}
              price={price}
              origin={origin}
              instructions={instructions || 'Not provided'}
              isClickAndCollect={isClickAndCollect}
            />
          </div>
        </React.Fragment>
      ) : (
        <div className={cn('loading')}>
          <Skeleton height={25} count={2} />
        </div>
      )}
    </div>
  );
};

DeliveryDetails.propTypes = propTypes;

export default DeliveryDetails;
