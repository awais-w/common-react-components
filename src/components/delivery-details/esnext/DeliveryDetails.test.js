import React from 'react';
import { shallow } from 'enzyme'; // eslint-disable-line import/no-extraneous-dependencies
import Skeleton from 'react-loading-skeleton';
import Pane from '@cmc/pane/esnext/pane';
import ContactDetails from '@cmc/contact-details/esnext';
import Address from '@cmc/address/esnext';
import DeliveryDetails from './DeliveryDetails';
import DeliveryTracking from './components/delivery-tracking';

describe('DeliveryDetails', () => {
  const setup = ({ contact, address, deliveryTracking }) => {
    const component = shallow(
      <DeliveryDetails contact={contact} address={address} deliveryTracking={deliveryTracking} />,
    );
    const block = 'cmc-delivery-details';
    return {
      wrapper: component,
      getPane: () => component.find(Pane),
      getContactDetailsCard: () =>
        component.findWhere((element) => element.name() === 'Card' && element.prop('title1').endsWith('Contact')),
      getDeliveryDetailsCard: () =>
        component.findWhere((element) => element.name() === 'Card' && element.prop('title1').endsWith('Info')),
      getContactDetails: () => component.find(`.${block}__contact-details`),
      getAddress: () => component.find(`.${block}__address`),
      getInstructions: () => component.find('.cmc-address'),
      getDeliveryTracking: () => component.find(`.${block}__delivery-tracking`),
      getSkeleton: () => component.find(Skeleton),
    };
  };
  const contact = {
    business: 'My business',
    person: { title: 'Mr', name: 'Francisco', surname: 'Lopez', middleName: 'Jose' },
    emails: [
      { type: 'type1', address: 'address1' },
      { type: 'type2', address: 'address2' },
    ],
    telephones: [
      { type: 'type1', number: 'number1' },
      { type: 'type2', number: 'number2' },
    ],
  };
  const address = {
    city: 'city',
    country: 'country',
    county: 'county',
    line1: 'line1',
    line2: 'line2',
    line3: 'line3',
    postcode: 'postcode',
  };
  const deliveryTracking = {
    bookedSlot: '11:00am - 12:00pm',
    price: 'Â£6.98',
    actualArrival: '7:46am',
    plannedArrival: '7:45am',
    dropNumber: 26,
    origin: '0932',
    instructions: 'instructions',
  };

  describe('renders', () => {
    const component = setup({ contact, address, deliveryTracking });
    it('exists', () => expect(component.wrapper).toExist());
    it('should match default snapshot', () => {
      expect(component.wrapper).toMatchSnapshot();
    });
  });

  describe('renders when no props provided', () => {
    const component = setup({});
    it('exists', () => expect(component.wrapper).toExist());
  });

  describe('content', () => {
    const component = setup({ contact, address, deliveryTracking });
    describe('contact details', () => {
      const contactDetailsBlock = component.getContactDetails();
      it('renders the block', () => {
        expect(contactDetailsBlock).toExist();
      });
      it('renders contact details', () => {
        expect(contactDetailsBlock.containsMatchingElement(<ContactDetails {...contact} />)).toBe(true);
      });
    });
    describe('address', () => {
      const contactDetailsBlock = component.getContactDetails();
      it('renders the block', () => {
        expect(contactDetailsBlock).toExist();
      });
      it('renders the address', () => {
        expect(contactDetailsBlock.containsMatchingElement(<Address address={address} mode={'INLINE'} />)).toBe(true);
      });
    });
    describe('delivery tracking', () => {
      const deliveryTrackingBlock = component.getDeliveryTracking();
      it('renders the block', () => {
        expect(deliveryTrackingBlock).toExist();
      });
      it('renders the delivery tracking', () => {
        expect(deliveryTrackingBlock.containsMatchingElement(<DeliveryTracking {...deliveryTracking} />)).toBe(true);
      });
      it('if instructions are not provided, passes instructions prop with value "Not provided"', () => {
        const withoutInstructions = { ...deliveryTracking, instructions: '' };
        const componentWithoutInstructions = setup({ contact, address, deliveryTracking: withoutInstructions });
        const deliveryTrackingComp = componentWithoutInstructions.getDeliveryTracking();
        expect(
          deliveryTrackingComp.containsMatchingElement(
            <DeliveryTracking {...deliveryTracking} instructions='Not provided' />,
          ),
        ).toBe(true);
      });
    });
  });

  describe('when isLoading', () => {
    const component = setup({});
    it('renders a skeleton', () => {
      component.wrapper.setProps({ isLoading: true });
      expect(component.getSkeleton().exists()).toBeTruthy();
    });
  });
});
