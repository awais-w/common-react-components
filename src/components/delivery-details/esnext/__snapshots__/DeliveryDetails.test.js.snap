// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeliveryDetails renders should match default snapshot 1`] = `
<div
  className="cmc-delivery-details"
>
  <div
    className="cmc-delivery-details__contact-details"
  >
    <ContactDetails
      business="My business"
      emails={
        Array [
          Object {
            "address": "address1",
            "type": "type1",
          },
          Object {
            "address": "address2",
            "type": "type2",
          },
        ]
      }
      groupFields={true}
      person={
        Object {
          "middleName": "Jose",
          "name": "Francisco",
          "surname": "Lopez",
          "title": "Mr",
        }
      }
      telephones={
        Array [
          Object {
            "number": "number1",
            "type": "type1",
          },
          Object {
            "number": "number2",
            "type": "type2",
          },
        ]
      }
    />
    <Address
      address={
        Object {
          "city": "city",
          "country": "country",
          "county": "county",
          "line1": "line1",
          "line2": "line2",
          "line3": "line3",
          "postcode": "postcode",
        }
      }
      mode="INLINE"
    />
  </div>
  <div
    className="cmc-delivery-details__delivery-tracking"
  >
    <DeliveryTracking
      actualArrival="7:46am"
      bookedSlot="11:00am - 12:00pm"
      dropNumber={26}
      instructions="instructions"
      origin="0932"
      plannedArrival="7:45am"
      price="Â£6.98"
    />
  </div>
</div>
`;
